server/service/microsoft_mdm.go

```diff
@@ -1248,17 +1249,125 @@ func (svc *Service) isTrustedRequest(ctx context.Context, reqSyncML *fleet.SyncM
 	return errors.New("calling device is not trusted")
 }
 
+// isValidUPN checks if the provided user ID is a valid UPN
+func isValidUPN(userID string) bool {
+	const upnRegex = `^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$`
+	re := regexp.MustCompile(upnRegex)
+	return re.MatchString(userID)
+}
+
+// isFleetdPresentOnDevice checks if the device requires Fleetd to be deployed
+func (svc *Service) isFleetdPresentOnDevice(ctx context.Context, deviceID string) (bool, error) {
+	// checking first if the device was enrolled through programmatic flow
+	enrolledDevice, err := svc.ds.MDMWindowsGetEnrolledDeviceWithDeviceID(ctx, deviceID)
+
+	if err != nil || enrolledDevice == nil {
+		return false, errors.New("device not found")
+	}
```

Comment:
It will return either a value or an error, so you can simplify the error handling to this:
```suggestion
	if err != nil {
		return false, ctxerr.Wrap(ctx, err, "get windows enrolled device")
	}
```